Implement coherence delta vs. date plot

nackerley opened the pull request:
### Implements
This pull request implements:
- #4 
- #8 

## Outstanding
- Deprecate global variables:
> - I thought I saw a global variable used, but in fact this is only the case the slider and intensity plots are un-commented. That, said there were "redefinition of global variable" warnings which I've scrubbed out.
> - At a certain point I was thinking that when we get to the point of being able to switch between rasters as described in #3 and between polygon overlays as per #2 it may be necessary to record the state of the relevant radio buttons to determine what to display, when different points are selected on the coherence plot, and that the way to do this is using dcc.Store. But now I'm thinking it's probably possible to query the state of the radio buttons directly.
> - Conclusion: I don't foresee any need to share data between components.
- #11
- #12

nackerley commented:
I see that some checks were not successful, but I'm pretty sure the errors are all in the setup, not in the code I've pushed.

drotheram commented on .gitignore:
```diff
@@ -1,4 +1,3 @@
 virtualenv/*
-.vscode/settings.json
-config.ini
```
Config.ini should remain in the .gitignore file

nackerley replied:
Why? I don't understand.

My motivation in including it was that the system didn't work "out of the box" for lack of a config.ini, yet there was a sample_config.ini.

drotheram replied:
We should add a note in the Readme for the user to copy `sample_config.ini` to `config.ini` the contents of config.ini are only relevant to users with access to our cloud resources.

nackerley replied:
Agreed that if we're not going to include `config.ini` in the repo it should documented. But it seems simpler to just include it in the repo - what's the downside?

drotheram replied:
Listing our service endpoints, even if secure, could potentially be a security target

nackerley replied:
Done.

drotheram commented on .vscode/launch.json:
```diff
@@ -0,0 +1,25 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
```
Can we remove the IntelliSense plug?

nackerley replied:
Absolutely!

nackerley commented:
Any progress on getting pylint to pass?

drotheram commented on app/dash_app.py:
```diff
@@ -1,383 +1,397 @@
-    dates = dates.replace('-', '').replace('"', '')
-    print(dates)
-    return 'cite:{}.adf.wrp.geo'.format(dates)
+        # html.Div(
```
Let's go ahead and remove this commented code. It is in the git history if we ever need it

drotheram commented on app/dash_app.py:
```diff
@@ -1,383 +1,397 @@
+                y=subset.index,
+                hovertemplate=(
+                    'Second: %{x}<br>'
+                    'First: -%{y} days<br>'
```
Would like to see both dates on the hover in addition to time between dates. This is useful but there are many scenarios where the analyst will want to know if the interferogram displayed covers a specific event or date. Datetime module should be able to calculate this easily with what's already here. Let's keep the temporal baseline as an additional piece of info. The integer doesn't need to be negative.

nackerley replied:
Agreed that showing the date would be better - it's going to take a bit of work to implement that, though. I don't think it's going to be possible to calculate it on the fly, I think you need to supply the extra data using the `customdata` attribute as described at https://plotly.com/python/hover-text-and-formatting/.

The negative sign is to indicate that the first date is always before (less than) the second. I thought it was clearer than just giving the delta.

nackerley replied:
Added issue #14 to be resolved by a future pull request.

drotheram commented on app/dash_app.py:
```diff
@@ -1,383 +1,397 @@
 if __name__ == '__main__':
-    app.run_server(host='0.0.0.0', port=8050, debug=False)
+    # TODO login and set up - or at least test - port forwarding
+    # See https://shorturl.at/lnSY1
```
Remove ToDo comments and create repo issue

nackerley replied:
Issue exists #12. We can remove the TODO when the issue is cleared.

drotheram commented on app/dash_app.py:
```diff
@@ -1,383 +1,397 @@
+config = get_config_params('config.ini')
+GEOSERVER_ENDPOINT = config.get('geoserver', 'geoserverEndpoint')
+
+# TODO add support for some or all of the following parameters to config
```
remove 'ToDo' and create issue

nackerley replied:
Issue exists #11 and we can remove TODO when it is closed.

drotheram commented on app/dash_app.py:
```diff
@@ -1,383 +1,397 @@
+CMAP_NAME = 'RdYlBu_r'
+COH_LIMS = (0.25, 0.65)
+
+# TODO read target configuration from database
```
remove todo and create repo issue

nackerley replied:
Issue #11 exists, and TODO can be removed when it is resolved.

drotheram reviewed (changes requested):
This is a good pass at adding the coherence timedelta. I like the added functionality of the hover. Before we merge I am requesting some changes. 

- Coherence timedelta plot gaps. This is probably going to be unsustainable for any of our beams that get bumped seasonally each new year will require a new plot gap
- Map and plots are too tall everything should be visible without scrolling. I had to zoom out to 75% zoom for the dashboard not to appear clausterphobic. This is probably because I assigned pixel heights rather than % heights initially. With the added plot below the screen gets too crowded.
- Coherence dot colouring scale is too wide. Anything above 0.3 should have a very hot colour, between 0.2 - 0.3 should still be relatively hot. With this stretch 0.3 coherence value interferograms are shades of blue
- Y-axis of coherence plot should probably be named 'Temporal Baseline (days)' as standard InSAR phraseology

drotheram commented:
Re: Coherence timedelta plot gaps. My suggestion is to limit visible scale to a fixed temporal baseline, such as 60-80 days. and allow the user to zoom in/out, pan, or zoom to the full network. This visualization needs to support the general case of coherence networks over long periods and different sensors with different repeat pass intervals. We will also likely see different seasonal trends if we begin to examine volcano hazards outside of Canada. 

RE: "Remember that part of keeping this usable is going to be "clearing out" date-pairs that turn out to be useless, like summer compared to winter" - Agreed, we can discard interferograms below a certain coherence threshold but once they have been processed and an average coherence value has been generated, we might as well retain that value so that we know empirically that the coherence was calculated and it was low.

nackerley commented:
@drotheram, I have fixed the CI/CD linters. We now have a PEP8 check which must pass, and a requirement that the pylint score be high (>9). See https://github.com/Volcano-Risk-Reduction-in-Canada/Volcanic-Interpretation-Workbench/tree/feature/coherence_timedelta/.github/workflows and the green checkmarks above.

I think I've addressed all the changes you've requested.
